phi_four.py

```diff
@@ -75,7 +73,7 @@ def forward(self, phi_state: torch.Tensor) -> torch.Tensor:
         action = (
             (2+0.5*self.m_sq)*phi_state**2 + # phi^2 terms
             self.lam*phi_state**4 - #phi^4 term
```

Comment:
I've tried to implement the optimisation
\sum_x phi(x) ( phi(x-mu) + phi(x+mu) ) = 2 \sum_x phi(x) phi(x+mu)
where I've shifted x -> x+mu in the first term.
The idea is that if you are summing pairs of neighbours over the whole lattice, the order doesn't matter.
In other words, when you plug eq (21) into eq (20) you see that the two terms with shifts in eq (21) give actually the same contribution to the action.
The action in (20) is unchanged: the difference is that instead of doing 0.5*(shift_forward_term + shift_backward_term) we can just do simply 0.5\*2\*shift_forward_term=shift_forward_term.
I tried a minimal modification, hoping not to spoil the shift functions... but I don't exclude that in the code there are other places where having backward and forward shifts is assumed.
You can also try to compute the action of a configuration with both formulations and check that the results are the same (it works for me for L=2,4).

Reply:
The other factors look ok to me (but maybe it is easier to check having our action in latex...)

Reply:
Added some notes on this and improved documentation after our discussion